### ISSUE TITLE
Support form validation and more data types in questions.yaml

### ISSUE THREAD

**Issue by gaktive** [rancher/dashboard#5754]
New types to add:

- [x] annotations - gives you a key-value pair component
- [x] labels - just like annotations except it doesn't allow multiline strings
- [x] references to other objects including secrets and configmaps
- [x] NodeSelectors
- [x] tolerations
- [x] AffinityRules
- [x] list of strings
- [ ] toleration array (specifically requested here https://github.com/rancher/dashboard/issues/2703)
- [ ] inline file yaml
- [ ] inline file IMI
- [ ] cloud credentials
- [ ] radio buttons
- [ ] validate that the checkbox already works
- [ ] can leverage values.schema.json https://github.com/rancher/dashboard/issues/2158. Example json schema uses this format: https://json-schema.org/learn/examples/calendar.schema.json

Additional Questions.yaml validation options to support (forward porting from Ember to dashboard):

- [ ] `min_length`
- [ ] `max_length`
- [ ] `min`
- [ ] `max`
- [ ] `valid_chars`
- [ ] `invalid_chars`
- [ ] `required`

If one of the fields is invalid, you should see an error message and you shouldn't be able to install the Helm chart.

Source: Question variable reference https://rancher.com/docs/rancher/v2.6/en/helm-charts/creating-apps/#question-variable-reference

**Comment by gaktive** [comment]
This is tied to the questions.yaml which @Sean-McQ has discussed before.

**Comment by catherineluse** [comment]
You're right, it's probably better to put it in 2.7 because even if we got it in by 2.6.7, it would still need backend changes (adding a new questions.yaml in the Helm chart repo) in order to affect the user experience for monitoring or project monitoring.

**Comment by catherineluse** [comment]
We could add this to the other enhancements for questions.yaml https://github.com/rancher/dashboard/issues/5754

**Comment by xhejtman** [comment]
with `required` attribute, it could be problem if form is split to several "tabs", user can have some difficulties to search which tab is missing required value.

**Comment by staedter** [comment]
Is there anyway to validate the questions.yaml. I modified mine for a cluster template chart but know I somehow broke it and now the ui form is not even shown and its just an empty yaml editor when I try to create a new cluster with this template.

**Comment by lindhe** [comment]
I agree with @staedter ! It would be so helpful if there was a `schema.json` to validate `questions.yaml`!

**Comment by lindhe** [comment]
It's worth noting that while `required` does not prevent the user from clicking "Install", it does at least provide a little * in the GUI...

![image](https://github.com/rancher/dashboard/assets/7773090/28a43c14-d92f-4a6f-9fd1-4b95ca079bf5)

**Comment by richard-cox** [comment]
The description needs updating given https://github.com/rancher/dashboard/issues/9588

**Comment by ThoSap** [comment]
As @richard-cox mentioned in the last comment, this issue description needs to be updated as his issue #9588 and his related PR #9488 fixes most of the missing checkbox items in Rancher v2.8.0.
This also adds support for **tooltips**, which is new.

The documentation was not updated yet since Rancher v2.8.0 as the issue rancher/rancher-docs#1051 is not addressed yet.

**New/enhanced/existing types that are not checked in the issue description above**
For all supported types see https://github.com/rancher/dashboard/blob/master/shell/components/Questions/index.vue#L20-L41

- [x] searchable and multi-choice enums (new using a YAML map / `multiple: true`)
- [x] inline file yaml
- [x] cloud credentials
- [x] radio buttons
- [x] boolean (checkbox)
- [x] hostname
- [x] ipaddr
- [x] cidr
- [x] cron

**Additional Questions.yaml validation options**
Reimplemented again in the new dashboard: https://github.com/rancher/dashboard/blob/master/shell/utils/validators/index.js#L96-L131

- [x] min
- [x] max
- [x] min_length
- [x] max_length
- [x] valid_chars
- [x] invalid_chars
- [x] required

**Comment by ThoSap** [comment]
If someone searches for a workaround for `valid_chars` to support a full regex and not only a regex character set, see https://github.com/rancher/rancher-docs/issues/847#issuecomment-2419687167 😉
